### ISSUE TITLE
Improve configurability

### ISSUE THREAD

**Issue by badosu** [vhakulinen/gnvim#59]
At the moment, configuration requires changing the `.desktop` file or running gnvim as a command.

It would be ideal to keep as much configuration as possible as vim global variables or respecting vim options, but if not feasible (e.g. for externalized features) use a xdg config file.

**Comment by smolck** [comment]
I haven't looked into this, but [neovim-gtk has some decent-looking configuration](https://github.com/daa84/neovim-gtk/wiki/Configuration) (see [here](https://github.com/daa84/neovim-gtk/wiki/Configuration#options) for the way it handles disabling external features). Not sure if we want to go the same route. It might be nice to have a simple option of `g:gnvim_external_tabline` or something that could be set to 0 or 1 in `init.vim` or similar (although we might want something different for on-the-fly configuring).

**Comment by vhakulinen** [comment]
Looks like `neovim-gtk` is just calling [set_option](https://github.com/daa84/neovim-gtk/blob/e4482a679963177d2affb91131636e872dfe6c14/src/nvim/redraw_handler.rs#L150), so doing the same on gnvim shouldn't be too hard.

**Comment by smolck** [comment]
How do we want to go about exposing the configuration options to the user (instead of command line args)? Once I know I can probably start working on a PR for this (I imagine most if not all of the refactoring will go into `main.rs`?).

**Comment by vhakulinen** [comment]
Messages over rpc might be the only way to do this currently. Thats how the ["gnvim" events](https://github.com/vhakulinen/gnvim/blob/5bf01b9b965d580355b7abc66419dee1c93d1c58/src/nvim_bridge/mod.rs#L909) are done currently.

**Comment by smolck** [comment]
@vhakulinen @badosu Okay, so with that info I have a (fairly small!) working diff, let me know if this is worth opening a PR for (right now you have to manually `:call rpcnotify(g:gnvim_channel_id, 'Gnvim', 'DisableExtTabline')`; I assume we'll want to create a commands.vim file or similar for it):
```diff
diff --git a/src/nvim_bridge/mod.rs b/src/nvim_bridge/mod.rs
index 2a723fc..06e2cd8 100644
--- a/src/nvim_bridge/mod.rs
+++ b/src/nvim_bridge/mod.rs
@@ -424,6 +424,7 @@ impl fmt::Display for RedrawEvent {
 #[derive(Debug, PartialEq)]
 pub enum GnvimEvent {
     SetGuiColors(SetGuiColors),
+    DisableExtTabline,
     CompletionMenuToggleInfo,
 
     CursorTooltipLoadStyle(String),
@@ -911,6 +912,7 @@ pub(crate) fn parse_gnvim_event(
 ) -> Result<GnvimEvent, String> {
     let cmd = try_str!(args.get(0).ok_or("No command given")?, "cmd");
     let res = match cmd {
+        "DisableExtTabline" => GnvimEvent::DisableExtTabline,
         "SetGuiColors" => {
             let mut colors = SetGuiColors::default();
 
diff --git a/src/ui/ui.rs b/src/ui/ui.rs
index 6285d46..41303d1 100644
--- a/src/ui/ui.rs
+++ b/src/ui/ui.rs
@@ -7,7 +7,7 @@ use std::time;
 use gdk;
 use glib;
 use gtk;
-use neovim_lib::neovim::Neovim;
+use neovim_lib::neovim::{Neovim, UiOption};
 use neovim_lib::neovim_api::NeovimApi;
 use neovim_lib::NeovimApiAsync;
 use neovim_lib::Value;
@@ -435,6 +435,13 @@ fn handle_gnvim_event(
     nvim: Arc<Mutex<Neovim>>,
 ) {
     match event {
+        GnvimEvent::DisableExtTabline => {
+            let mut nvim = nvim.lock().unwrap();
+            if let Err(err) = nvim.set_option(UiOption::ExtTabline(false)) {
+                nvim.command_async(&format!(
+                        "echo \"Failed to disable ext_tabline: '{}'\"", err));
+            }
+        }
         GnvimEvent::SetGuiColors(colors) => {
             state.popupmenu.set_colors(colors.pmenu, &state.hl_defs);
             state.tabline.set_colors(colors.tabline, &state.hl_defs);
```
Let me know what you think.

**Comment by badosu** [comment]
Nice! It seems that's the way to go as delineated by @vhakulinen

**Comment by smolck** [comment]
Cool! Will get a PR for it shortly (if all goes well).

**Comment by smolck** [comment]
@badosu @vhakulinen What do you guys think of making GNvim configurable (in the future when it gets more features and options) via JSON? It may not be ideal, but there are some upsides:

- It's (relatively) easy to parse before starting GNvim (via [`serde_json`](https://github.com/serde-rs/json)).
- It's simple to change/edit through a GUI (e.g. "settings" GUI that outputs the JSON options into a config file)
- It would not require a whole lot of boilerplate inside of GNvim
- It's an easily understandable format, so no need to understand/learn YAML, VimL, Lua, etc. to configure GNvim (makes GNvim config more accessible IMHO)
- It could be used for simple configuration of both GNvim and also external plugins written in Python/Lua/etc., e.g. the following:
    ```json
    {
        "pyScriptedPlugin.option": true,
        "disableExtTabline": false,
        "gnvimTitle": "GNvim - STUFF"
    }
    ```

While this would stray from the "stay true to Neovim" idea in a sense, I think JSON does have the upside of (possibly) being easier to deal with as config options grow (among other upsides like those I stated above). Note that GNvim would still need a way to be extensible via plugins written in Lua/Python/VimL/whatever; a JSON settings file would just act as a place for any plugins (and GNvim itself) to put configuration options.

As a side note, I've been using VSCode (Electron . . . yikes!) a bit and that's where I've gotten this idea from.

**Comment by vhakulinen** [comment]
What benefits would that bring over compared to viml?

**Comment by vhakulinen** [comment]
Everything in gnvim should be accessible through viml.

**Comment by smolck** [comment]
> Everything in gnvim should be accessible through viml.

Makes sense; it is GN*vim* after all.

> Besides, what would you configure through the json config file?

Well, anything, really. Stuff like GNvim’s title, font, what external features are disabled, etc. TBH, now that I think about it, it doesn’t make much sense for GNvim to use JSON; since we have vimL, it’d be redundant. Sticking to vimL definitely makes more sense; thanks for pointing that out.
